### PR TITLE
Random Number Generator

### LINKED ISSUE #128: feat: Random number generation
In advanced concepts chapter, with security concerns.

### PR DESCRIPTION
**Pull request by 0xNeshi**
**Issue(s):**  Close #128 

### Description

Built off of https://github.com/NethermindEth/StarknetByExample/pull/183

Explain randomness in the context of blockchain, and Starknet specifically, and provide an example with `CoinFlip` contract.

### Checklist

- [x] **CI Verifier:** Run `./scripts/cairo_programs_verifier.sh` successfully
- [x] **Contract Tests:** Added tests to cover the changes
- [x] **Deployed Contract (Optional):** If possible, include a [Voyager link to a deployed contract on Sepolia](https://sepolia.voyager.online/contract/0x040783d486a1bbc3727b1177b7dec6aa139c1e6265be4788a70938ecd2eaa017)
    - [fund the contract with ETH](https://sepolia.voyager.online/contract/0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7#writeContract) if necessary:

### CODE REVIEW

File: listings/applications/coin_flip/src/tests.cairo [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+    let randomness_dispatcher = IRandomnessDispatcher { contract_address: randomness_address };
+    let coin_flip_dispathcer = ICoinFlipDispatcher { contract_address: coin_flip_address };
+
+    (coin_flip_dispathcer, randomness_dispatcher, eth_dispatcher, deployer)
+}
+
+#[test]
+#[fuzzer(runs: 10, seed: 22)]
+fn test_two_flips(random_word_1: felt252, random_word_2: felt252) {
```
Comment (misicnenad) [inline listings/applications/coin_flip/src/tests.cairo:56]: Created this test bearing in mind that I wanted to test 2 things at the same time:
1. Flipping a coin multiple times
2. Flipping a coin with random values (fuzz testing)

File: src/SUMMARY.md [hunk 2]
```diff
@@ -62,6 +62,7 @@ Summary
 - [Simple Storage with Starknet-js](./applications/simple_storage_starknetjs.md)
 - [Crowdfunding Campaign](./applications/crowdfunding.md)
 - [AdvancedFactory: Crowdfunding](./applications/advanced_factory.md)
+- [Random Number Generator](./applications/random_number_generator.md)
```
Comment (misicnenad) [inline src/SUMMARY.md:65]: Wasn't sure about the title to use here, open to suggestions

**Comment by misicnenad** [comment]
Converting to draft: fixing the fee refund logic

**Comment by misicnenad** [comment]
> Converting to draft: fixing the fee refund logic

In the end simplified the contract logic to make the example focus more on the _randomness generation_ part.

**Review by julio4 — CHANGES_REQUESTED** [review]
LGTM! Really good tests and nice to mock the randomness part as well.
Some small changes/comments and it'll be good!

File: src/applications/random_number_generator.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+- **Description:** Allowing users to provide entropy directly, often combined with other sources to generate a random number.
+- **Example:** Users submitting their own random values which are then hashed together with other inputs.
+- **Risks:**
+    - **Collusion:** Users may collude to provide inputs that skew the randomness in their favor.
+    - **Front-Running:** Other participants might observe a user's input and act on it before it gets included in the block, affecting the outcome.
+
+#### 3. External Oracles
+
+- **Description:** Using a trusted third-party service to supply randomness. Oracles are off-chain services that provide data to smart contracts.
+- **Example:** Chainlink VRF (Verifiable Random Function) is a service that provides cryptographically secure randomness.
```
[the remark this answers is not in the document]
Comment (misicnenad) [reply src/applications/random_number_generator.md:35]: Meant to write "Pragma" here, brain fart

File: listings/applications/coin_flip/src/contract.cairo [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        pub flip_id: u64,
+        pub flipper: ContractAddress,
+        pub side: Side
+    }
+
+    #[derive(Drop, Debug, PartialEq, Serde)]
+    pub enum Side {
+        Heads,
+        Tails,
+        Sideways
```
[the remark this answers is not in the document]
Comment (misicnenad) [reply listings/applications/coin_flip/src/contract.cairo:59]: Not really, I included it just for fun, will remove

File: listings/applications/coin_flip/src/contract.cairo [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+#[starknet::contract]
+pub mod CoinFlip {
+    use core::num::traits::zero::Zero;
+    use starknet::{ContractAddress, get_caller_address, get_contract_address,};
+    use pragma_lib::abi::{IRandomnessDispatcher, IRandomnessDispatcherTrait};
+    use openzeppelin::token::erc20::interface::{IERC20Dispatcher, IERC20DispatcherTrait};
+
+    #[storage]
+    struct Storage {
+        eth_dispatcher: IERC20Dispatcher,
```
[the remark this answers is not in the document]
Comment (misicnenad) [reply listings/applications/coin_flip/src/contract.cairo:29]: Is there a downside to storing dispatchers as opposed to addresses?
Seems like re-creating the dispatcher on each call makes for more expensive operations :thinking: 

Chose _not_ to hardcode the ETH address, as this allows me to inject a mock ERC20 contract address during testing

File: src/applications/random_number_generator.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    - **Collusion:** If participants collude, they can influence the randomness outcome.
+    - **Disruption:** If participants do not follow through (e.g., not revealing their values), the beacon may fail to produce a valid random number.
+
+## CoinFlip using Pragma VRF
+
+Below is an implementation of a `CoinFlip` contract that utilizes a [Pragma Verifiable Random Function (VRF)](https://docs.pragma.build/Resources/Cairo%201/randomness/randomness) to generate random numbers on-chain.
+
+```rust
+{{#include ../../listings/applications/coin_flip/src/contract.cairo}}
+```
```
[the remark this answers is not in the document]
Comment (misicnenad) [reply src/applications/random_number_generator.md]: Let me know what you think https://github.com/NethermindEth/StarknetByExample/pull/238/commits/5c68e776b9e9736e2251e5e2fccde50e098fd2be

**Review by julio4 — APPROVED** [review]

**Review by julio4 — APPROVED** [review]
Thank you!

**Pull request merged**